src/CMakeLists.txt

```diff
@@ -120,10 +124,22 @@ IF (PAHO_BUILD_STATIC)
                 ${CMAKE_BINARY_DIR})
     ENDFOREACH()
 
-    INSTALL(TARGETS paho-mqtt3c-static paho-mqtt3a-static
-        ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR})
+    IF (NOT PAHO_BUILD_SHARED)
```

Comment:
Is it correct, that static libraries will only be installed, if no shared libraries are build?

Reply:
I would prefer, in general, that it be left up to the user whether to build and install the static libraries, shared libraries, or both. Certainly CMake allows optimization of the build in the case of both by using object libraries so that the sources are only compiled once and then linked into the two libs.

With both static and shared libs installed, it's then up to each application or project to decide which lib is better for the project. (As I juggle multiple projects/customers simultaneously, this is essential).

I will look at this in detail later today, but in the meantime, here is an example of what I did for another library, albeit with my limited CMake skills:
https://github.com/fpagliughi/sockpp/blob/master/CMakeLists.txt

Reply:
I believe I understand what you are saying. However, building both libraries at the same time presents a difficulty. At least for windows, it is not simply a link-line choice for the consumer between the static and the shared libraries. To make the _both_ option work, the end consumer must set a `#define` at compile time to specify whether they desire to link against the static or shared version of the library in addition to specifying the correct library on the link-line.

This is because when a header declares the following function:

`int MQTTAsync_reconnect(MQTTAsync handle);`

...after preprocessor macros are evaluated, the code will either evaluate to:

`__declspec(dllimport) int MQTTAsync_reconnect(MQTTAsync handle);`

or 

`int MQTTAsync_reconnect(MQTTAsync handle);`

but it cannot be both at the same time. You need an additional `#define` from the user to specify which one of the two declarations is desired by the user _at compile time_. The first tells the compiler to expect the DLL import library on the link line, the second works for the static library.

The second line (without the __declspec) technically works for both the static and DLL libraries. However, according to Microsoft documentation, specifying the dllimport causes the compiler to generate more efficient code. Just removing the dllimport was the approach I took on my first submission, but that was undesired.

Making it build one way or the other simplifies the CMake files quite a bit. Additionally, there shouldn't actually be any end-user that should actually _need_ both configurations at the same time.